{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e8048e94",
   "metadata": {},
   "source": [
    "fig4B \n",
    "\n",
    "trajectory analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46a8b246",
   "metadata": {},
   "outputs": [],
   "source": [
    "# monocle2\n",
    "library(monocle)\n",
    "library(Seurat)\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "library(viridis)\n",
    "args <- commandArgs(T)\n",
    "\n",
    "my_color = c(\n",
    "    \"0\"=\"#93d0fc\", \"1\"=\"#2aaae1\", \"2\"=\"#3f5246\", \"3\"=\"#ad3074\", \"4\"=\"#E7B10A\", \n",
    "    \"5\"=\"#91a3b0\", \"6\"=\"#d58282\", \"7\"=\"#418557\", \"8\"=\"#6dc831\", \"9\"=\"#eacd01\",\n",
    "    \"10\"=\"#ff9f22\", \"11\"=\"#af5b2b\", \"12\"=\"#2e5073\", \"13\"=\"#E6D2AA\", \"14\"=\"#42032C\",\n",
    "    \"15\"=\"#752f9a\", \"16\"=\"#721c28\", \"17\"=\"#2c2cff\", \"18\"=\"#d1bac8\", \"19\"=\"#009580\",\n",
    "    \"20\"=\"#94ac78\", \"21\"=\"#9f59f8\", \"22\"=\"#ffff00\", \"23\"=\"#fc0cd4\", \"24\"=\"#ffccff\"\n",
    ")\n",
    "\n",
    "potato_merged=readRDS(args[1])\n",
    "outprefix <- args[2]\n",
    "\n",
    "potato_merged$seurat_clusters <- as.character(potato_merged$seurat_clusters)\n",
    "Idents(potato_merged) <- potato_merged$seurat_clusters\n",
    "\n",
    "if (!dir.exists(outprefix)){\n",
    "  dir.create(outprefix)\n",
    "}\n",
    "\n",
    "run_monocle2=function(sbj){\n",
    "    expr_matrix = as(as.matrix(sbj@assays$Spatial@counts),'sparseMatrix')\n",
    "    p_data=sbj@meta.data\n",
    "    p_data$celltype=sbj@active.ident # 当前活动的idents changed\n",
    "    f_data=data.frame(gene_short_name=row.names(sbj),row.names=row.names(sbj))\n",
    "    pd=new('AnnotatedDataFrame',data=p_data)\n",
    "    fd=new('AnnotatedDataFrame',data=f_data)\n",
    "    cds=newCellDataSet(expr_matrix,\n",
    "                    phenoData=pd,\n",
    "                    featureData=fd,\n",
    "                    lowerDetectionLimit=0.5,\n",
    "                    expressionFamily=negbinomial.size())\n",
    "\n",
    "    cds <- estimateSizeFactors(cds)\n",
    "    cds <- estimateDispersions(cds)\n",
    "    return(cds)\n",
    "}\n",
    "\n",
    "GM_state <- function(cds,root){\n",
    "    if (length(unique(pData(cds)$State)) > 1){\n",
    "        T0_counts <- table(pData(cds)$State,pData(cds)$celltype)[,root]\n",
    "        return(as.numeric(names(T0_counts)[which(T0_counts == max(T0_counts))]))\n",
    "    } else {\n",
    "        return (1)\n",
    "    }\n",
    "}\n",
    "\n",
    "run_seuratDEG=function(sbj,cds,out_name,root){\n",
    "    sbj <- NormalizeData(sbj, normalization.method = \"LogNormalize\", scale.factor = 10000)\n",
    "    sbj <- ScaleData(sbj) \n",
    "    deg.cluster <- FindAllMarkers(sbj)\n",
    "    diff.genes <- subset(deg.cluster,p_val<0.05)$gene\n",
    "    cds <- setOrderingFilter(cds, diff.genes)\n",
    "    cds=reduceDimension(cds,max_components=2,method=\"DDRTree\")\n",
    "    cds=orderCells(cds)\n",
    "    root_in=GM_state(cds, root)\n",
    "    cds=orderCells(cds,root_state=root_in)\n",
    "\n",
    "    p=plot_cell_trajectory(cds,color_by=\"Pseudotime\",size=1,show_backbone=TRUE,cell_size=2) + scale_color_gradientn(colours =c(viridis(20)))\n",
    "    ggsave(filename=paste0(out_name,\".monocle.pseudotime.seuratDEG.pdf\"),width=14,height=14)\n",
    "    p=plot_cell_trajectory(cds,color_by=\"State\",size=1,show_backbone=TRUE)\n",
    "    ggsave(filename=paste0(out_name,\".monocle.state.seuratDEG.pdf\"),width=14,height=14)\n",
    "    p=plot_cell_trajectory(cds,color_by=\"celltype\",size=1,show_backbone=TRUE) + scale_color_manual(values=my_color)\n",
    "    ggsave(filename=paste0(out_name,\".monocle.celltype.seuratDEG.pdf\"),width=14,height=14)\n",
    "    p=plot_cell_trajectory(cds,color_by=\"inf_stat\",size=1,show_backbone=TRUE)\n",
    "    ggsave(filename=paste0(out_name,\".monocle.inf_stat.seuratDEG.pdf\"),width=14,height=14)\n",
    "    \n",
    "    cds\n",
    "}\n",
    "\n",
    "# all cells\n",
    "print(\"all cells\")\n",
    "cds <- run_monocle2(potato_merged)\n",
    "\n",
    "cds <- run_seuratDEG(potato_merged, cds, paste0(outprefix, \"potato_merged.inf2noninf\"), \"16\")\n",
    "saveRDS(cds, \"potato_merged.monocle2.seuratDEG.pdf.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79d93537",
   "metadata": {},
   "outputs": [],
   "source": [
    "# monocle3\n",
    "library(Seurat)\n",
    "library(monocle3)\n",
    "set.seed(12345)\n",
    "\n",
    "args <- commandArgs(T)\n",
    "data_set <- args[1]\n",
    "outprefix <- args[2]\n",
    "if (!dir.exists(outprefix)){\n",
    "  dir.create(outprefix)\n",
    "}\n",
    "\n",
    "potato_merged=readRDS(data_set)\n",
    "potato_merged$seurat_clusters <- as.character(potato_merged$seurat_clusters)\n",
    "Idents(potato_merged) <- potato_merged$seurat_clusters\n",
    "\n",
    "dat_part <- potato_merged\n",
    "\n",
    "run_monocle3 <- function(sbj){\n",
    "    expr_matrix = as(as.matrix(sbj@assays$Spatial@counts),'sparseMatrix')\n",
    "    p_data=sbj@meta.data\n",
    "    p_data$celltype=sbj@active.ident # 聚类的结果\n",
    "    f_data=data.frame(gene_short_name=row.names(sbj),row.names=row.names(sbj))\n",
    "    pd=new('AnnotatedDataFrame',data=p_data)\n",
    "    fd=new('AnnotatedDataFrame',data=f_data)\n",
    "    cds=new_cell_data_set(expr_matrix,\n",
    "                          cell_metadata = p_data,\n",
    "                          gene_metadata = f_data)\n",
    "    return(cds)\n",
    "}\n",
    "\n",
    "get_earliest_principal_node <- function(cds, root){\n",
    "  cell_ids <- which(colData(cds)[, \"seurat_clusters\"] == root) # 使用idents的列\n",
    "  \n",
    "  closest_vertex <-\n",
    "  cds@principal_graph_aux[[\"UMAP\"]]$pr_graph_cell_proj_closest_vertex\n",
    "  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])\n",
    "  root_pr_nodes <-\n",
    "  igraph::V(principal_graph(cds)[[\"UMAP\"]])$name[as.numeric(names\n",
    "  (which.max(table(closest_vertex[cell_ids,]))))]\n",
    "  \n",
    "  root_pr_nodes\n",
    "}\n",
    "\n",
    "# all cells\n",
    "cds <- run_monocle3(dat_part)\n",
    "\n",
    "reducedDim(cds, type = \"PCA\") <- dat_part@reductions$pca@cell.embeddings\n",
    "cds@preprocess_aux$prop_var_expl <- dat_part@reductions$pca@stdev\n",
    "\n",
    "# umap in cell.embeddings\n",
    "# cds@int_colData@listData$reducedDims$UMAP <- dat_part@reductions$umap@cell.embeddings\n",
    "# umap in meta.data\n",
    "umap.tab <- data.frame(UMAP_1=dat_part$UMAP_1, UMAP_2=dat_part$UMAP_2)\n",
    "rownames(umap.tab) <- rownames(dat_part@meta.data)\n",
    "cds@int_colData@listData$reducedDims$UMAP <- umap.tab\n",
    "cds@clusters$UMAP_so$clusters <- Idents(dat_part)\n",
    "cds <- cluster_cells(cds, reduction_method = \"UMAP\", resolution = 1e-3)\n",
    "\n",
    "# must set to NULL so it could continue learn graph \n",
    "rownames(cds@principal_graph_aux$UMAP$dp_mst) <- NULL\n",
    "colnames(cds@int_colData@listData$reducedDims$UMAP) <- NULL\n",
    "# calculate presudotime\n",
    "cds <- learn_graph(cds, use_partition = T)\n",
    "cds <- order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds, '16'))\n",
    "cds$seurat_clusters <- as.character(cds$seurat_clusters)\n",
    "\n",
    "write.table(pData(cds), paste0(outprefix, \".monocle3.pData.table\"), col.names=T, row.names=F, sep=\"\\t\", quote=F)\n",
    "\n",
    "pdf(paste0(outprefix, \".monocle3.celltype.pdf\"))\n",
    "plot_cells(cds,\n",
    "           color_cells_by = \"inf_stat\",\n",
    "           label_cell_groups=FALSE,\n",
    "           label_leaves=FALSE,\n",
    "           label_branch_points=FALSE,\n",
    "           graph_label_size=1.5,\n",
    "           show_trajectory_graph=F)\n",
    "dev.off()\n",
    "pdf(paste0(outprefix, \".monocle3.seurat_clts.pdf\"))\n",
    "plot_cells(cds,\n",
    "           color_cells_by = \"seurat_clusters\",\n",
    "           label_cell_groups=FALSE,\n",
    "           label_leaves=FALSE,\n",
    "           label_branch_points=FALSE,\n",
    "           graph_label_size=1.5,\n",
    "           show_trajectory_graph=F)\n",
    "dev.off()\n",
    "pdf(paste0(outprefix, \".monocle3.pseudotime.pdf\"))\n",
    "plot_cells(cds,\n",
    "           color_cells_by = \"pseudotime\",\n",
    "           label_cell_groups=FALSE,\n",
    "           label_leaves=FALSE,\n",
    "           label_branch_points=FALSE,\n",
    "           graph_label_size=1.5,\n",
    "           show_trajectory_graph=F)\n",
    "dev.off()\n",
    "pdf(paste0(outprefix, \".monocle3.state.pdf\"))\n",
    "plot_cells(cds,\n",
    "           color_cells_by = \"cluster\",\n",
    "           label_cell_groups=FALSE,\n",
    "           label_leaves=FALSE,\n",
    "           label_branch_points=FALSE,\n",
    "           graph_label_size=1.5,\n",
    "           show_trajectory_graph=F)\n",
    "dev.off()\n",
    "\n",
    "saveRDS(cds, file=paste0(outprefix, \".monocle3.rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80bc9ed8",
   "metadata": {},
   "source": [
    "fig4D\n",
    "\n",
    "pseudo-time heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6f0af38",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEG analysis\n",
    "\n",
    "library(Seurat)\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "\n",
    "potato_merged=readRDS('final_used_data.RDS')\n",
    "input_clt <- c(16, 6, 10, 11, 8, 7, 19, 2)\n",
    "dat_part <- subset(potato_merged, subset=(inf_stat==\"CK\")|((inf_stat!=\"CK\")&(seurat_clusters %in% input_clt)))\n",
    "dat_part$DEG <- ifelse(dat_part$inf_stat==\"CK\", \"CK\", dat_part$seurat_clusters)\n",
    "#dat_part$DEG <- ifelse((dat_part$seurat_clusters==16 | dat_part$seurat_clusters==6)&dat_part$inf_stat!=\"CK\", \"inf\", dat_part$DEG)\n",
    "#dat_part$DEG <- ifelse((dat_part$seurat_clusters==10 | dat_part$seurat_clusters==11 | dat_part$seurat_clusters==8)&dat_part$inf_stat!=\"CK\", \"mid\", dat_part$DEG)\n",
    "#dat_part$DEG <- ifelse((dat_part$seurat_clusters==7 | dat_part$seurat_clusters==19 | dat_part$seurat_clusters==2)&dat_part$inf_stat!=\"CK\", \"noninf\", dat_part$DEG)\n",
    "Idents(dat_part) <- dat_part$DEG\n",
    "#Idents(dat_part) <- dat_part$seurat_clusters\n",
    "\n",
    "dat.part.3h <- subset(dat_part, subset=(hours==\"3T\"))\n",
    "dat.part.6h <- subset(dat_part, subset=(hours==\"6h\"))\n",
    "dat.part.12h <- subset(dat_part, subset=(hours==\"12T\"))\n",
    "dat.part.24h <- subset(dat_part, subset=(hours==\"24T\"))\n",
    "\n",
    "diff_exp_paired <- function(obj, id1, id2, outprefix){\n",
    "    diff.wilcox = FindMarkers(obj, only.pos = FALSE, \n",
    "                                min.pct = 0.25, logfc.threshold = 0.1,\n",
    "                                ident.1 = id1, ident.2 = id2\n",
    "                                )\n",
    "    diff.wilcox$gene = rownames(diff.wilcox)\n",
    "    all.markers = diff.wilcox %>% select(gene, everything()) %>% subset(p_val<0.05)\n",
    "    all.markers$gene=paste0('A157_',all.markers$gene)\n",
    "    write.table(all.markers,\n",
    "    file=paste0(outprefix, \".DEG_gene.txt\"),\n",
    "    sep=\"\\t\",col.names=T,quote=F,row.names=F)\n",
    "}\n",
    "diff_exp <- function(obj, outprefix){\n",
    "    diff.wilcox = FindAllMarkers(obj, only.pos = T, \n",
    "                                min.pct = 0.25, logfc.threshold = 0.1\n",
    "                                )\n",
    "    all.markers = diff.wilcox %>% select(gene, everything()) %>% subset(p_val<0.05)\n",
    "    all.markers$gene=paste0('A157_',all.markers$gene)\n",
    "    write.table(all.markers,\n",
    "    file=paste0(outprefix, \".marker_gene.txt\"),\n",
    "    sep=\"\\t\",col.names=T,quote=F,row.names=F)\n",
    "}\n",
    "diff_exp(dat.part.3h, \"3h.inf_mid_noninf_cltmrk\")\n",
    "diff_exp(dat.part.6h, \"6h.inf_mid_noninf_cltmrk\")\n",
    "diff_exp(dat.part.12h, \"12h.inf_mid_noninf_cltmrk\")\n",
    "diff_exp(dat.part.24h, \"24h.inf_mid_noninf_cltmrk\")\n",
    "diff_exp(dat_part, \"all.inf_mid_noninf_cltmrk\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b3b2061",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this script used to add cell distribution upon heatmap \n",
    "# eg. 12h cluster8 inf noninf cells upon pesudotime split \n",
    "\n",
    "library(monocle)\n",
    "library(Seurat)\n",
    "library(tidyverse)\n",
    "\n",
    "my_color = c(\n",
    "      \"0\"=\"#93d0fc\", \"1\"=\"#2aaae1\", \"2\"=\"#3f5246\", \"3\"=\"#ad3074\", \"4\"=\"#E7B10A\", \n",
    "      \"5\"=\"#91a3b0\", \"6\"=\"#d58282\", \"7\"=\"#418557\", \"8\"=\"#6dc831\", \"9\"=\"#eacd01\",\n",
    "      \"10\"=\"#ff9f22\", \"11\"=\"#af5b2b\", \"12\"=\"#2e5073\", \"13\"=\"#E6D2AA\", \"14\"=\"#42032C\",\n",
    "      \"15\"=\"#752f9a\", \"16\"=\"#721c28\", \"17\"=\"#2c2cff\", \"18\"=\"#d1bac8\", \"19\"=\"#009580\",\n",
    "      \"20\"=\"#94ac78\", \"21\"=\"#9f59f8\", \"22\"=\"#ffff00\", \"23\"=\"#fc0cd4\", \"24\"=\"#ffccff\"\n",
    "  )\n",
    "\n",
    "args <- commandArgs(T)\n",
    "\n",
    "potato_merged=readRDS(args[1])\n",
    "potato_merged$DEG <- ifelse(potato_merged$seurat_clusters==16 | potato_merged$seurat_clusters==6, \"inf\", \"N\")\n",
    "potato_merged$DEG <- ifelse(potato_merged$seurat_clusters==10 | potato_merged$seurat_clusters==11 | potato_merged$seurat_clusters==8, \"mid\", potato_merged$DEG)\n",
    "potato_merged$DEG <- ifelse(potato_merged$seurat_clusters==7 | potato_merged$seurat_clusters==19 | potato_merged$seurat_clusters==2, \"noninf\", potato_merged$DEG)\n",
    "Idents(potato_merged) <- potato_merged$DEG\n",
    "potato_merged$seurat_clusters <- as.character(potato_merged$seurat_clusters)\n",
    "gene_f <- args[2]\n",
    "\n",
    "run_monocle2=function(sbj){\n",
    "    expr_matrix = as(as.matrix(sbj@assays$Spatial@counts),'sparseMatrix')\n",
    "    p_data=sbj@meta.data\n",
    "    p_data$celltype=sbj@active.ident\n",
    "    f_data=data.frame(gene_short_name=row.names(sbj),row.names=row.names(sbj))\n",
    "    pd=new('AnnotatedDataFrame',data=p_data)\n",
    "    fd=new('AnnotatedDataFrame',data=f_data)\n",
    "    cds=newCellDataSet(expr_matrix,\n",
    "                    phenoData=pd,\n",
    "                    featureData=fd,\n",
    "                    lowerDetectionLimit=0.5,\n",
    "                    expressionFamily=negbinomial.size())\n",
    "\n",
    "    cds <- estimateSizeFactors(cds)\n",
    "    cds <- estimateDispersions(cds)\n",
    "    return(cds)\n",
    "}\n",
    "\n",
    "cds <- run_monocle2(potato_merged)\n",
    "\n",
    "GM_state <- function(cds,root){\n",
    "    if (length(unique(pData(cds)$State)) > 1){\n",
    "        T0_counts <- table(pData(cds)$State,pData(cds)$seurat_clusters)[,root]\n",
    "        return(as.numeric(names(T0_counts)[which(T0_counts == max(T0_counts))]))\n",
    "    } else {\n",
    "        return (1)\n",
    "    }\n",
    "}\n",
    "\n",
    "run_seuratDEG=function(sbj,cds,out_name,root){\n",
    "    sbj <- NormalizeData(sbj, normalization.method = \"LogNormalize\", scale.factor = 10000)\n",
    "    sbj <- ScaleData(sbj) \n",
    "    deg.cluster <- FindAllMarkers(sbj)\n",
    "    diff.genes <- subset(deg.cluster,p_val<0.05)$gene\n",
    "    cds <- setOrderingFilter(cds, diff.genes)\n",
    "    cds=reduceDimension(cds,max_components=2,method=\"DDRTree\")\n",
    "    cds=orderCells(cds)\n",
    "    root_in=GM_state(cds, root)\n",
    "    cds=orderCells(cds,root_state=root_in)\n",
    "\n",
    "    p=plot_cell_trajectory(cds,color_by=\"Pseudotime\",size=1,show_backbone=TRUE,cell_size=2)\n",
    "    ggsave(filename=paste0(out_name,\".monocle.pseudotime.seuratDEG.pdf\"),width=14,height=14)\n",
    "    p=plot_cell_trajectory(cds,color_by=\"State\",size=1,show_backbone=TRUE)\n",
    "    ggsave(filename=paste0(out_name,\".monocle.state.seuratDEG.pdf\"),width=14,height=14)\n",
    "    p=plot_cell_trajectory(cds,color_by=\"celltype\",size=1,show_backbone=TRUE)\n",
    "    ggsave(filename=paste0(out_name,\".monocle.celltype.seuratDEG.pdf\"),width=14,height=14)\n",
    "    p=plot_cell_trajectory(cds,color_by=\"seurat_clusters\",size=1,show_backbone=TRUE)\n",
    "    ggsave(filename=paste0(out_name,\".monocle.seurat_clts.seuratDEG.pdf\"),width=14,height=14)\n",
    "\n",
    "    cds\n",
    "}\n",
    "\n",
    "cds <- run_seuratDEG(potato_merged, cds, \"inf_mid_noninf\", \"16\")\n",
    "\n",
    "# immune gene innner time\n",
    "immune_gene <- read.table(gene_f, sep=\"\\t\", header = T)[[\"gene\"]] %>% gsub(\"A157_\", \"\", .) %>% unique(.)\n",
    "\n",
    "# plot pesudo time heatmap\n",
    "p <- plot_pseudotime_heatmap(cds[immune_gene,], \n",
    "num_clusters = 10, \n",
    "cores = 10, \n",
    "show_rownames = F, \n",
    "return_heatmap=T)\n",
    "\n",
    "p_order <- p$tree_row$order\n",
    "p_label <- p$tree_row$label\n",
    "p_label <- p_label[p_order] # reorder the label\n",
    "genecluster <- as.data.frame(cutree(p$tree_row, k = 10))\n",
    "colnames(genecluster) <- c(\"cluster\")\n",
    "genecluster$gene <- rownames(genecluster)\n",
    "genecluster <- genecluster[p_label,]\n",
    "write.table(genecluster, file=paste0(\"inf_mid_noninf\", \".pesudo_order_20231026.tab\"),\n",
    "col.names = T, row.names = F, sep = \"\\t\", quote = F)\n",
    "\n",
    "saveRDS(p, file=paste0(\"inf_mid_noninf\", \".pheatmap.pesudo_order_20231026.rds\"))\n",
    "pdf(paste0(\"inf_mid_noninf\", \".pheatmap.pesudo_order_20231026.pdf\"), width = 7.5, height = 7.5)\n",
    "grid::grid.newpage()\n",
    "grid::grid.draw(p$gtable)\n",
    "dev.off() \n",
    "\n",
    "# prepare data\n",
    "newdata <- seq(min(pData(cds)$Pseudotime), max(pData(cds)$Pseudotime),length.out = 100)\n",
    "cell_pseudo_df <- data.frame(Pseudotime = pData(cds)$Pseudotime, celltype = pData(cds)$seurat_clusters)\n",
    "\n",
    "sep <- (newdata[2] - newdata[1])/2\n",
    "plot_df = cell_pseudo_df[(cell_pseudo_df$Pseudotime >= newdata[1] & cell_pseudo_df$Pseudotime < newdata[1] + sep ), ]\n",
    "plot_df = as.data.frame(table(plot_df$celltype))\n",
    "# plot_df$celltype = factor(plot_df$celltype) 必须为因子类型\n",
    "plot_df$group <- \"box1\"\n",
    "\n",
    "for (i in seq(2,100)){\n",
    "    if (i == 100){\n",
    "        tmp_df = cell_pseudo_df[((cell_pseudo_df$Pseudotime >= newdata[i]-sep) & (cell_pseudo_df$Pseudotime <= newdata[i])), ]\n",
    "        tmp_df = as.data.frame(table(tmp_df$celltype))\n",
    "        tmp_df$group <- paste0(\"box\", as.character(i))\n",
    "        plot_df <- rbind(plot_df, tmp_df)\n",
    "    }else {\n",
    "        tmp_df = cell_pseudo_df[((cell_pseudo_df$Pseudotime >= newdata[i]-sep) & (cell_pseudo_df$Pseudotime < newdata[i] + sep)), ]\n",
    "        tmp_df = as.data.frame(table(tmp_df$celltype))\n",
    "        tmp_df$group <- paste0(\"box\", as.character(i))\n",
    "        plot_df <- rbind(plot_df, tmp_df)\n",
    "    }\n",
    "}\n",
    "\n",
    "# 山脊图\n",
    "library(ggridges)\n",
    "\n",
    "df_list <- list()\n",
    "for (clt in names(table(plot_df$Var1))){\n",
    "    tmp_df <- plot_df[plot_df$Var1==clt, ]\n",
    "    tmp_df <- tmp_df %>% mutate(expanded = map2(group, Freq, rep)) %>%\n",
    "    unnest(expanded)\n",
    "    tmp_df <- as.data.frame(tmp_df)\n",
    "    tmp_df$expanded <- as.integer(gsub(\"box\", \"\", tmp_df$expanded))\n",
    "    df_list[[clt]] <- tmp_df\n",
    "}\n",
    "combined_df <- do.call(rbind, df_list)\n",
    "\n",
    "p <- ggplot(combined_df, aes(x=expanded, y = Var1, fill = Var1)) +\n",
    "geom_density_ridges() +\n",
    "theme_ridges() +\n",
    "scale_fill_manual(values = my_color) +\n",
    "theme(legend.position = \"none\") + scale_x_continuous(breaks=seq(0, 100, 20))\n",
    "\n",
    "ggsave(file=paste0(\"inf_mid_noninf\", \".ggridges.pesudo_order_20231102.pdf\"), plot=p, width = 5, height = 1.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61395640",
   "metadata": {},
   "source": [
    "fig4e\n",
    "\n",
    "key TF identified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a05acde",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "set.seed(12345)\n",
    "library(\"dplyr\")\n",
    "library(\"Seurat\")\n",
    "library(\"future\")\n",
    "library(tidyverse)\n",
    "library(Matrix)\n",
    "options(future.globals.maxSize = 240000 * 1024^2)\n",
    "\n",
    "# diff exp paired\n",
    "diff_exp_paired <- function(obj, id1, id2, outprefix){\n",
    "    diff.wilcox = FindMarkers(obj, only.pos = F, \n",
    "                                min.pct = 0.25, logfc.threshold = 0.1,\n",
    "                                ident.1 = id1, ident.2 = id2\n",
    "                                )\n",
    "    diff.wilcox$gene = rownames(diff.wilcox)\n",
    "    all.markers = diff.wilcox %>% select(gene, everything()) %>% subset(p_val_adj<0.05)\n",
    "    all.markers$gene=paste0('A157_',all.markers$gene)\n",
    "    write.table(all.markers,\n",
    "    file=paste0(outprefix, \"_marker_gene.txt\"),\n",
    "    sep=\"\\t\",col.names=T,quote=F,row.names=F)\n",
    "}\n",
    "\n",
    "args <- commandArgs(T)\n",
    "clt1 <- args[1]\n",
    "clt2 <- args[2]\n",
    "\n",
    "obj <- readRDS(\"../Dataset2.6h.20231022.rds\")\n",
    "diff_exp_paired(obj, clt1, clt2, paste0(\"6h_\", as.character(clt1), \"_\",  as.character(clt2)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ec88f51",
   "metadata": {},
   "source": [
    "fig4E"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b95e960",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ComplexHeatmap)\n",
    "library(ggplot2)\n",
    "library(stringr)\n",
    "library(tidyverse)\n",
    "library(RColorBrewer)\n",
    "library(reshape2)\n",
    "library(viridis)\n",
    "\n",
    "args <- commandArgs(T)\n",
    "\n",
    "dat_part <- readRDS(args[1])\n",
    "gene.list <- read.table(args[2], header = T, sep = \"\\t\")[[\"gene\"]] %>% gsub(\"A157_\", \"\", .) %>% unique(.) %>% rev(.)\n",
    "\n",
    "dat_part$DEG2 <- factor(dat_part$DEG2, levels=c(\"16\", \"6\", \"10\", \"11\", \"8\", \"7\", \"2\"))\n",
    "Idents(dat_part) <- dat_part$DEG2\n",
    "dat_part <- subset(dat_part, subset=DEG2!=\"19\")\n",
    "plot_dotplot <- function(dat_part, gene.list, smax, ph, outputprefix){\n",
    "    p1 <- DotPlot(object = dat_part, features = gene.list, scale.min = 0, scale.max=smax) + \n",
    "    theme_classic() +\n",
    "    theme(panel.grid.major=element_blank(),panel.grid.minor=element_blank(),panel.background = element_blank()) +\n",
    "    theme(axis.text.x = element_text(hjust = 1, vjust = .5,angle = 90,size=7),\n",
    "    panel.background = element_rect(fill=\"white\"),\n",
    "    axis.title.x = element_text(colour = \"white\")) +\n",
    "    coord_flip() + \n",
    "    scale_colour_gradient2(low = \"#deebf7\", mid = \"#6baed6\", high = \"#08306b\")\n",
    "\n",
    "    ggsave(p1, filename=outputprefix, width=5, height=ph)\n",
    "}\n",
    "\n",
    "smax_h <- c(\n",
    "    \"3T\"=50,\n",
    "    \"6h\"=50,\n",
    "    \"12T\"=50,\n",
    "    \"24T\"=50\n",
    ")\n",
    "\n",
    "ph <- 7\n",
    "dat_part.tmp <- dat_part\n",
    "plot_dotplot(dat_part.tmp, gene.list, smax_h[['6h']], ph, gsub(\".txt\", paste0(\".20240423.dotplot.pdf\"), args[2]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae9ca3c6",
   "metadata": {},
   "source": [
    "fig4F\n",
    "\n",
    "pyscenic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3c9703f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################################\n",
    "# PART1\n",
    "# arboreto and GRNBoost2\n",
    "#########################################################\n",
    "# https://academic.oup.com/bioinformatics/article/35/12/2159/5184284\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from arboreto.algo import grnboost2\n",
    "from distributed import LocalCluster, Client\n",
    "\n",
    "import sys\n",
    "\n",
    "if __name__=='__main__':\n",
    "\n",
    "    local_cluster = LocalCluster(n_workers=10,\n",
    "                                threads_per_worker=1,\n",
    "                                memory_limit=8e9)\n",
    "    custom_client = Client(local_cluster)\n",
    "    \n",
    "    ex_matrix = pd.read_csv(sys.argv[1], sep='\\t', header=0, index_col=-1)\n",
    "    ex_matrix.shape\n",
    "    tf_df = pd.read_csv(sys.argv[2], sep=\"\\t\", header=0, index_col=None)\n",
    "    tf_df['tf'] = tf_df.apply(lambda x: x['gene'].replace(\"A157_\", \"\"), axis=1)\n",
    "    tf_list = list(tf_df['tf'])\n",
    "    \n",
    "    adjancencies = grnboost2(expression_data=ex_matrix, tf_names=tf_list, verbose=True, seed=12345)\n",
    "    adjancencies.to_csv(sys.argv[3], index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a366b7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################################\n",
    "# PART2\n",
    "# cisTarget\n",
    "#########################################################\n",
    "\n",
    "#########################################################\n",
    "# STEP1 motif meme -> cluster burst\n",
    "# python\n",
    "#########################################################\n",
    "\n",
    "from Bio import motifs\n",
    "\n",
    "motifs_list = []\n",
    "motifs_f = open(r\"Stu_TF_binding_motifs.meme\", \"r\")\n",
    "for m in motifs.parse(motifs_f, \"minimal\"):\n",
    "    motifs_list.append(m)\n",
    "print(motifs_list[0].counts)\n",
    "print(dir(motifs_list[0]))\n",
    "motifs_f.close()\n",
    "\n",
    "gene_dict = {}\n",
    "with open(r\"A157mapDM.txt\", \"r\") as map_f:\n",
    "    for line in map_f:\n",
    "        tmp = line.strip(\"\\n\").split(\"\\t\")\n",
    "        potato_g = tmp[0].split('.')[0]\n",
    "        DM_g = tmp[1]\n",
    "        if DM_g in gene_dict.keys():\n",
    "            if potato_g not in gene_dict[DM_g]:\n",
    "                gene_dict[DM_g].append(potato_g)\n",
    "            else:\n",
    "                pass\n",
    "        else:\n",
    "            gene_dict[DM_g] = [potato_g]\n",
    "            \n",
    "for m in motifs_list:\n",
    "    tmp_m = m\n",
    "    if m.name not in gene_dict.keys():\n",
    "        continue\n",
    "    for g in gene_dict[m.name]:\n",
    "        with open(g.replace(\"A157_\", \"\") + \".cb\", \"w\") as outf:\n",
    "            tmp_m.name = g\n",
    "            motif_cb = motifs.write([tmp_m], \"clusterbuster\")\n",
    "            outf.write(motif_cb)\n",
    "    \n",
    "#########################################################\n",
    "# STEP2 prepare cisTarget db\n",
    "# shell\n",
    "#########################################################\n",
    "\"\"\"\n",
    "ls Stu_TF_binding_motifs_cb | sed 's/\\.cb//g' > motifs.lst\n",
    "\n",
    "python /path/to/create_cisTarget_databases/create_cisTarget_databases-master/create_cistarget_motif_databases.py \\\n",
    "-f A157_franking_10kb.rename.fa \\\n",
    "-M Stu_TF_binding_motifs_cb \\\n",
    "-m motifs.lst \\\n",
    "-o A157 \\\n",
    "-t 20 \\\n",
    "-c cbust \\\n",
    "-g #[0-9]+$ # 必须 通过motifs id 获取gene id \n",
    "\"\"\"\n",
    "\n",
    "#########################################################\n",
    "# STEP3 motif ann\n",
    "# python\n",
    "#########################################################\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "flist = [f.replace(\".cb\", \"\") for f in os.listdir(\"Stu_TF_binding_motifs_cb\")]\n",
    "df = pd.DataFrame(\n",
    "    {\"#motif_id\": flist}\n",
    ")\n",
    "df[\"gene_name\"] = df[\"#motif_id\"]\n",
    "df[\"motif_similarity_qvalue\"] = 0\n",
    "df[\"orthologous_identity\"] = 1\n",
    "df[\"description\"] = df[\"#motif_id\"]\n",
    "\n",
    "df.to_csv(\"motif.tbl\", header=True, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e526257d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################################\n",
    "# PART3\n",
    "# run pyscenic\n",
    "#########################################################\n",
    "\n",
    "#########################################################\n",
    "# STEP1 prepare_ctx_input.py\n",
    "# python\n",
    "#########################################################\n",
    "import pandas as pd\n",
    "\n",
    "# exp mat\n",
    "EXP_MAT_FNAME = \"exp_mat.tab\"\n",
    "\n",
    "# RIBOSOMAL genes\n",
    "ribo_df = pd.read_csv(\"Ribo.txt\", header=0, index_col=None, sep=\"\\t\")\n",
    "RIBOSOMAL_GENE = list(ribo_df['gene'])\n",
    "\n",
    "# adjancencies matrix from GRNBoost2\n",
    "ADJACENCIES_FNAME = \"adjancencies.tsv\"\n",
    "ADJACENCIES_FILTED_FNAME = \"adjancencies_filted.tsv\"\n",
    "\n",
    "# motif from cisTarget\n",
    "DATABASES_GLOB = \"A157.genes_vs_motifs.rankings.feather\"\n",
    "MOTIF_ANNOTATIONS_FNAME=\"motif.tbl\"\n",
    "\n",
    "def edges_filter(adj_df):\n",
    "    # Derive potential regulomes from these co-expression modules \n",
    "    adj_sorted_df = adj_df.sort_values(['TF', 'importance'], ascending=[1,0])\n",
    "\n",
    "    # remove ribosomal\n",
    "    adj_sorted_filted_df = adj_sorted_df[(~adj_sorted_df['target'].isin(RIBOSOMAL_GENE))]\n",
    "\n",
    "    return adj_sorted_filted_df\n",
    "    \n",
    "# RIBOSOMAL\n",
    "RIBOSOMAL_GENE = []\n",
    "\n",
    "if __name__=='__main__':\n",
    "\n",
    "    ex_matrix = pd.read_csv(EXP_MAT_FNAME, sep='\\t', header=0, index_col=-1)\n",
    "    ex_matrix.to_csv(\"exp_mat.tsv\", sep=\"\\t\", header=True, index=True) # cell x genes 3000 x36000\n",
    "\n",
    "    adjacencies = pd.read_csv(ADJACENCIES_FNAME, header=0, index_col=None, sep='\\t')\n",
    "    assert adjacencies.shape[1] == 3, \"check adjacenies whether it is right\"\n",
    "    adj_sorted_filted_df = edges_filter(adjacencies)\n",
    "    adj_sorted_filted_df.columns = [\"TF\", \"target\", \"importance\"]\n",
    "    adj_sorted_filted_df.to_csv(ADJACENCIES_FILTED_FNAME, header=True, index=False, sep=\"\\t\")\n",
    "\n",
    "#########################################################\n",
    "# STEP2 run_ctx\n",
    "# shell\n",
    "#########################################################\n",
    "\n",
    "\"\"\"\n",
    "# 注意输入\n",
    "# 邻接矩阵，表达矩阵必须以csv或tsv结尾\n",
    "/path/to/pyscenic ctx adjancencies.tsv A157.genes_vs_motifs.rankings.feather \\\n",
    "--annotations_fname motif.tbl \\\n",
    "--expression_mtx_fname ./exp_mat.tsv \\\n",
    "--top_n_targets 50 \\ # 筛选邻接边的方法\n",
    "--num_workers 20 \\\n",
    "--nes_threshold 2 \\ # module聚类阈值\n",
    "--mode \"dask_multiprocessing\" \\\n",
    "--output reg.csv \\\n",
    "--no_pruning \\ # 不进行过滤\n",
    "--min_genes 3\n",
    "\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
